File: algorithms/levenshtein-distance/answers/levenshtein-distance.rb
class String
  def levenshtein_distance(other)
    return other.length if self.length == 0
    return self.length if other.length == 0

    cost = 0

    if self[-1] == other[-1]
      cost = 1
    end

    sub_left = self[0...-1].levenshtein_distance(other)
    sub_right = self.levenshtein_distance(other[0...-1])
    sub_both = self[0...-1].levenshtein_distance(other[0...-1])

    [sub_left + 1, sub_right + 1, sub_both + cost].min
  end
end

puts 'kitten'.levenshtein_distance('sittin') == 3

File: algorithms/levenshtein-distance/answers/levenshtein-distance.js
String.prototype.levenshteinDistance = function(other) {
  if (this.length === 0) return other.length;
  if (other.length === 0) return this.length;

  let cost;

  if (this[this.length - 1] === other[other.length - 1]) {
    cost = 0;
  } else {
    cost = 1;
  }

  const subLeft = this.slice(0, -1).levenshteinDistance(other);
  const subRight = this.levenshteinDistance(other.slice(0, -1));
  const subBoth = this.slice(0, -1).levenshteinDistance(other.slice(0, -1));

  return Math.min(subLeft + 1, subRight + 1, subBoth + cost);
}

console.log('kitten'.levenshteinDistance('sitting') === 3);
